articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+
+The interface definition of the [LoadNode](https://github.com/ros2/rcl_interfaces/blob/master/composition_interfaces/srv/LoadNode.srv) service could be extended with a field to specify the desired `QoS` of the loaded node.
+
+In that way, the user loading the node is responsible of passing the desired QoS, instead of the node container.
+The message to be sent could be loaded from a qos profile file.
+
+#### Proposed interfaces
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# composition_interfaces/srv/LoadNode.srv
+
+# Current request fields
+...
+# New field
+EndpointQoS[] requested_qos_profiles  # new field to allow loading user requested qos profiles
+---
+# Current response definition
+...
+```
+
+A `rcl_qos_loaded_profiles_t` can be constructed from the received message and passed to the node options of the node to be loaded.
+
+TBD: `QoS` message definition.
+
+### Support in launch files
+
+`launch_ros.actions.Node` and `launch_ros.actions.LoadComposableNode` actions could support a qos file argument in their constructors.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+## Explicitly allowing external configurability in entities
+
+Both node and init options could have a flag to ignore the new `--qos-file` argument.
+In a similar way, there could be an option for publishers/subscriptions/clients and services to allow side loading their qos or not.
+
+In nodes designed to be reused, it does make sense to allow side-loading the qos settings of all entities.
+To make this use case easier, the following options can be added:
+
+- `rcl_init_options_t` is extended with a `bool allow_side_loading_qos`.
+- The following enum is defined:
+  ```c
+  typedef enum rcl_side_load_qos_t {
+      RCL_SIDE_LOAD_QOS_DEFAULT;
+      RCL_SIDE_LOAD_QOS_ENABLED;
+      RCL_SIDE_LOAD_QOS_DISABLED;
+  } rcl_side_load_qos_t;
+  ```
+- A `rcl_side_load_qos_t allow_side_loading_qos` member is added to `rcl_node_options_t`, it will copy the behavior defined in the `rcl_init_options_t` of its parent context when the enum value is `RCL_SIDE_LOAD_QOS_DEFAULT`.
```

Comment:
How does this strategy play out when loading composable nodes? For example, if I load a composable node without setting the "side load" policy, is it inheriting the behavior from the context provided by the container?

I don't think we want a loaded component to inherit an ENABLED flag, since we've previously expressed interest in making configurable QoS an opt-in feature. I.e. I think the default should always be "disabled" (regardless of the parent context settings). Then, we don't need an enum and can just use a boolean.

Reply:
> How does this strategy play out when loading composable nodes? For example, if I load a composable node without setting the "side load" policy, is it inheriting the behavior from the context provided by the container?

Using `allow_side_loading_qos=false` in the container init options solves the problem.
That would actually be the default, and it's not going to be an external mechanism to change it (it can only be changed in code).
The loaded node can tweak its node options to allow side loading or not, the container is not going to be able to make that decision.

> I don't think we want a loaded component to inherit an ENABLED flag, since we've previously expressed interest in making configurable QoS an opt-in feature. I.e. I think the default should always be "disabled" (regardless of the parent context settings). Then, we don't need an enum and can just use a boolean.

I think that deleting the three states enum in node options make sense, but not for subscriptions/publishers/clients services.
The most common use case is to enable the feature in all the sub-entities of a node, thus why I have added this option.

This allows granularity, it's default opt-out, the author of the node has complete control over it, and it can easily be opted-in for the most common use case (enabling the feature in all the sub-entities of a node).
The first three points where the concerns raised, and this mechanism solves them.

Not having a way of enabling easily the feature for all subentities of a node will end up in a flag being passed to 10000000 entities.

Reply:
> The loaded node can tweak its node options to allow side loading or not, the container is not going to be able to make that decision.

Okay, sounds good. I just wanted to make sure that it's not possible, by default, for an integrator to change the QoS of a nodes entities.

> I think that deleting the three states enum in node options make sense, but not for subscriptions/publishers/clients services.
The most common use case is to enable the feature in all the sub-entities of a node, thus why I have added this option.

:+1: